File: server/src/main/java/com/cspinformatique/kubik/domain/product/service/ProductService.java
package com.cspinformatique.kubik.domain.product.service;

import java.util.Set;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import com.cspinformatique.kubik.domain.dilicom.model.Reference;
import com.cspinformatique.kubik.model.product.Product;
import com.cspinformatique.kubik.model.product.Supplier;

public interface ProductService {
	Product buildProductFromReference(Reference reference);

	Iterable<Product> findByEan13(String ean13);
	
	Product findByEan13AndSupplier(String ean13, Supplier supplier);
	
	Iterable<Product> findBySupplier(Supplier supplier);
	
	Page<Product> findAll(Pageable pageable);
	
	Product findOne(int id);
	
	Set<String> getProductIdsCache();
	
	void mergeProduct(Product sourceProduct, Product targetProduct);
	
	Product save(Product product);
	
	Page<Product> search(String query, Pageable pageable);
}

File: kubik/kubik-server/src/main/webapp/public/js/product/product-service.js
(function(){
	var PRODUCT_URL = "/product";
	
	angular
		.module("Kubik")
		.factory("productService", ProductService);
	
	function ProductService($http){
		return {
			findByEan13 : findByEan13,
			mergeProducts : mergeProducts
		};
		
		function findByEan13(ean13){
			return $http
				.get(PRODUCT_URL + "?ean13=" + ean13)
				.then(findByEan13Success);
			
			function findByEan13Success(response){
				return response.data;
			}
		}
		
		function mergeProducts(sourceProduct, targetProduct){
			return $http
				.get(PRODUCT_URL + "/" + sourceProduct.id + "?mergeProduct=&targetProductId=" + targetProduct.id);
		}
	}
})();